### DISCUSSION TITLE
abs() is funky

### DISCUSSION THREAD

**Discussion by zoharl3** [yalmip/YALMIP#986] [Possible bug or missing functionality] [answered]
Sometimes, using abs() or norm(., 1) causes infeasibility, e.g.:

```
x = sdpvar( n, 1, 'full' );
obj = norm( x, 1 );
```

However, doing it manually works fine:

```
t = sdpvar( n, 1 );
obj = sum( t );
con = [ dl <= t; -dl <= t ];
```

I think it happened both with gurobi for MILP and mosek for SOCP.
If you are interested, then I can try to come up with an example. But it's not a priority--less convenient, but I can do it manually.

**Comment by johanlofberg** [comment, marked as answer]
You would have to supply an example, as that is definitely not expected behavior and I've never seen such an issue before (norm(.,1) derives exactly the model you show)

btw using 'full' when defining a vector is redundant

**Reply by zoharl3** [reply]
Here, it's a linear program:

https://drive.google.com/file/d/1094fYAjAkzQQZcJQf0j_AH3GC_wP0Rc6/view

I believe yalmip formulation causes numerical issues for gurobi. In this case, mosek handles it fine.

Yes, I know that 'full' is redundant. However, since the default 'sym' is very unintuitive for me, and it caused me trouble in the past, I tend to keep it safe.

**Reply by zoharl3** [reply]
To elaborate why I consider not using 'full' a pitfall. If I take a shortcut, then sometimes I'm tempted to write things like
```
x = sdpvar(n, 2); % UV coordinates of a mesh with n vertices
x = sdpvar(n, 3); % spatial coord.
x = sdpvar(n, 4); % n 2x2 matrices--jacobian of mapping a triangle
```
which is usually safe since I usually work with a large n. But sometimes, it's not true for a small model--by model I refer to a 3D mesh object. This surprise and oversight can cost me a day of debugging.

I mean the off chance that n happens to be equal to my const in the second dimension is quite slim. I wouldn't have chosen 'sym' as the default behavior for it. But I guess, it's too late for changing that at this point.

**Reply by johanlofberg** [reply]
This looks like a presolve bug in gurobi. Turn it off and there are no problems. Problem is reproduced if you switch order of your constraints to -dl <= t; dl <= t;  

Yes 'full' should be used when one argument isn't always 1.

**Reply by zoharl3** [reply]
It's not the first time gurobi's pre-solve gives me grief. Need to start watching it.
Thanks

**Reply by Derrickzc** [reply]
> This looks like a presolve bug in gurobi. Turn it off and there are no problems. Problem is reproduced if you switch order of your constraints to -dl <= t; dl <= t;
> 
> Yes 'full' should be used when one argument isn't always 1.

How to turn off presolve  in yalmip？ Can u tell me ?

**Reply by johanlofberg** [reply]
You change gurobi presolve options
https://yalmip.github.io/command/sdpsettings/

**Reply by Derrickzc** [reply]
> You change gurobi presolve options https://yalmip.github.io/command/sdpsettings/


I only learn how to close in guorbi process. 

%%params.Presolve=0; %% close presolve
%%rst = gurobi(model,params);

But in yalmip, I think I should use something like:
ops = sdpsettings ('verbose', 1, 'solver', 'gurobi', 'params.Presolve', '0')

**Reply by Derrickzc** [reply]
> You change gurobi presolve options https://yalmip.github.io/command/sdpsettings/

Or like 
ops = sdpsettings ('verbose', 1, 'solver', 'gurobi', 'gurobi.presolve', '0')?

**Reply by johanlofberg** [reply]
no according to the link it would be `'gurobi.presolve',0`

**Reply by Derrickzc** [reply]
> > You change gurobi presolve options https://yalmip.github.io/command/sdpsettings/
> 
> Or like ops = sdpsettings ('verbose', 1, 'solver', 'gurobi', 'gurobi.presolve', '0')?

In case anyone find this answer, plus
ops = sdpsettings ('verbose', 1, 'solver', 'gurobi', 'gurobi.MIPGap', 1e-3)?

**Comment by zoharl3** [comment]
An advice for future reference:

https://support.gurobi.com/hc/en-us/community/posts/4407464930321-pre-solve-is-buggy
